odiglet/pkg/kube/runtime_details/pods_controller.go

```diff
@@ -0,0 +1,153 @@
+// 1. When a new workload generation is applied, the runtime details might be changed (different env, versions, etc).
+// 2. When a source is added, but there are no running pods yet. When the first pod starts running, this is chance to apply runtime details detection.
+func (p *PodsReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := log.FromContext(ctx)
+
+	var pod corev1.Pod
+	err := p.Client.Get(ctx, request.NamespacedName, &pod)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	podWorkload, err := p.getPodWorkloadObject(ctx, &pod)
+	if err != nil {
+		logger.Error(err, "error getting pod workload object")
+		return reconcile.Result{}, err
+	}
+	if podWorkload == nil {
+		// pod is not managed by a workload, no runtime details detection needed
+		return reconcile.Result{}, nil
+	}
+
+	// get instrumentation config for the pod to check if it is instrumented or not
+	instrumentationConfigName := workload.CalculateWorkloadRuntimeObjectName(podWorkload.Name, podWorkload.Kind)
+	instrumentationConfig := odigosv1.InstrumentationConfig{}
+	err = p.Client.Get(ctx, client.ObjectKey{Name: instrumentationConfigName, Namespace: podWorkload.Namespace}, &instrumentationConfig)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	podGeneration, err := GetPodGeneration(ctx, p.Clientset, &pod)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// prevent runtime inspection on pods for which we already have the runtime details for this generation
+	if podGeneration == 0 || podGeneration <= instrumentationConfig.Status.ObservedWorkloadGeneration {
+		logger.Info("skipping redundant runtime details detection since generation is not newer", "name", request.Name, "namespace", request.Namespace, "currentPodGeneration", podGeneration, "observedWorkloadGeneration", instrumentationConfig.Status.ObservedWorkloadGeneration)
+		return reconcile.Result{}, nil
+	}
+
+	// The reconcile should be called every time the pod is updated and changes from not running to running
+
+	odigosConfig, err := k8sutils.GetCurrentOdigosConfig(ctx, p.Client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	runtimeResults, err := runtimeInspection([]corev1.Pod{pod}, odigosConfig.IgnoredContainers)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// persist the runtime results into the status of the instrumentation config
+	patchStatus := odigosv1.InstrumentationConfig{
+		Status: odigosv1.InstrumentationConfigStatus{
+			RuntimeDetailsByContainer:  runtimeResults,
+			ObservedWorkloadGeneration: podGeneration,
+		},
+	}
+	patchData, err := json.Marshal(patchStatus)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	err = p.Client.Status().Patch(ctx, &instrumentationConfig, client.RawPatch(types.MergePatchType, patchData), client.FieldOwner("odiglet-runtimedetails"))
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
```

Comment:
This part is common to this and the instrumentationConfigReconciler, right? maybe worth creating a function for it (`PersistRuntimeInspection`)?

Reply:
good idea, updated